calico.go

```diff
@@ -44,80 +45,93 @@ func init() {
 	// since namespace ops (unshare, setns) are done for a single thread, we
 	// must ensure that the goroutine does not jump from OS thread to thread
 	runtime.LockOSThread()
-
-	nodename, _ = os.Hostname()
-}
-
-func updateNodename(conf NetConf, logger *log.Entry) {
-	if conf.Hostname != "" {
-		nodename = conf.Hostname
-		logger.Warn("Configuration option 'hostname' is deprecated, use 'nodename' instead.")
-	}
-	if conf.Nodename != "" {
-		nodename = conf.Nodename
-	}
 }
 
 func cmdAdd(args *skel.CmdArgs) error {
 	// Unmarshal the network config, and perform validation
-	conf := NetConf{}
+	conf := types.NetConf{}
 	if err := json.Unmarshal(args.StdinData, &conf); err != nil {
 		return fmt.Errorf("failed to load netconf: %v", err)
 	}
 
-	cniVersion := conf.CNIVersion
+	utils.ConfigureLogging(conf.LogLevel)
 
-	ConfigureLogging(conf.LogLevel)
+	// Allow the nodename to be overridden by the network config
+	nodename = utils.DetermineNodename(conf)
 
-	workload, orchestrator, err := GetIdentifiers(args)
+	epIDs, err := utils.GetIdentifiers(args, nodename)
 	if err != nil {
 		return err
 	}
 
-	logger := CreateContextLogger(workload)
-
-	// Allow the nodename to be overridden by the network config
-	updateNodename(conf, logger)
-
-	logger.WithFields(log.Fields{
-		"Orchestrator": orchestrator,
-		"Node":         nodename,
-		"Workload":     workload,
-		"ContainerID":  args.ContainerID,
-	}).Info("Extracted identifiers")
-
+	logger := &log.Entry{}
+	if epIDs.OrchestratorID == "k8s" {
+		logger = log.WithFields(log.Fields{
+			"Workload":    epIDs.WEPName,
+			"ContainerID": epIDs.ContainerID,
+			"Pod":         epIDs.PodName,
+			"Namespace":   epIDs.Namespace,
+		})
+	} else {
+		logger = log.WithFields(log.Fields{
+			"Workload":    epIDs.WEPName,
+			"ContainerID": epIDs.ContainerID,
+		})
+	}
+	logger.Info("Extracted identifiers")
 	logger.WithFields(log.Fields{"NetConfg": conf}).Info("Loaded CNI NetConf")
-	calicoClient, err := CreateClient(conf)
+
+	calicoClient, err := utils.CreateClient(conf)
 	if err != nil {
 		return err
 	}
 
-	// Always check if there's an existing endpoint.
-	endpoints, err := calicoClient.WorkloadEndpoints().List(api.WorkloadEndpointMetadata{
-		Node:         nodename,
-		Orchestrator: orchestrator,
-		Workload:     workload})
+	ctx := context.Background()
+
+	// Check if there's an existing endpoint.
+	endpoints, err := calicoClient.WorkloadEndpoints().List(ctx, options.ListOptions{Name: epIDs.WEPName, Namespace: epIDs.Namespace, Prefix: true})
 	if err != nil {
 		return err
 	}
 
-	logger.Debugf("Retrieved endpoints: %v", endpoints)
+	endpoint := &api.WorkloadEndpoint{}
+	ids := names.WorkloadEndpointIdentifiers{
```

Comment:
I think this logic might be a little off.

IIUC, you want to zero-out the endpoint before you do the List and the NameMatches check - that way you're doing a prefix list without the endpoint name - so basically enumerating all endpoints under the workload (in the old hierarchical sense).

So rather than copying this struct I think just remove the endpoint and store off the old value for later.

Reply:
So, I don't think you need to copy this struct like this (and beside it feels like a bad thing to have to maintain)

Reply:
Oh - this is a completely different WorkloadEndpointsIdentifiers struct.  Hmmm - I think that's not necessary (see comment below).

Reply:
And if we aren't re-using the endpoint, why did we need to enumerate it?